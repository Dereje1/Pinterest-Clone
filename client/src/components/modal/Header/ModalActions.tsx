import React from 'react';
import IconButton from '@mui/material/IconButton';
import PushPinIcon from '@mui/icons-material/PushPin';
import PushPinOutlinedIcon from '@mui/icons-material/PushPinOutlined';
import DeleteForeverIcon from '@mui/icons-material/DeleteForever';
import Tooltip from '@mui/material/Tooltip';
import { PinType } from '../../../interfaces';

interface GetActionProps{
  element: PinType
  reset: (_: React.SyntheticEvent, forceClose?: boolean) => void
  pinImage: (pin: PinType) => void
  deletePin: ((pin: PinType) => void) | null
}

function GetAction({
  element,
  pinImage,
  deletePin,
  reset,
}: GetActionProps) {
  if (deletePin) { // means called from profile page
    return (
      <Tooltip title={element.owns ? 'Permanently delete image' : 'Unpin image'} placement="bottom">
        <IconButton
          aria-label="settings"
          style={{ margin: '1vh' }}
          onClick={(e) => {
            deletePin(element);
            reset(e);
          }}
          onMouseDown={(e) => e.preventDefault()}
        >
          {
            element.owns
              ? <DeleteForeverIcon style={{ fontSize: '2em', color: '#d12929' }} />
              : <PushPinIcon style={{ fontSize: '2em', color: '#3a1c1cde' }} />
          }
        </IconButton>
      </Tooltip>
    );
  }

  // finds the status of image to determine what kind of pin to place on image
  if (element.hasSaved || element.owns) { // If the user has already saved this pin
    return (
      <Tooltip title={element.owns ? 'You own this image' : 'Unpin this image'} placement="bottom">
        <IconButton
          aria-label="settings"
          style={{ margin: '1vh' }}
          disableRipple={element.owns}
          onClick={() => {
            if (element.hasSaved) {
              pinImage(element);
            }
          }}
          onMouseDown={(e) => e.preventDefault()}
        >

          <PushPinIcon style={{ fontSize: '2em', color: element.owns ? '#c50000' : '#3a1c1cde' }} />
        </IconButton>
      </Tooltip>
    ); // no button
  }
  // user has not saved this pin show outlined pin
  return (
    <Tooltip title="Pin image" placement="bottom">
      <IconButton
        aria-label="settings"
        style={{ margin: '1vh' }}
        onClick={() => {
          pinImage(element);
        }}
        onMouseDown={(e) => e.preventDefault()}
      >

        <PushPinOutlinedIcon style={{ fontSize: '2em' }} />
      </IconButton>
    </Tooltip>
  );
}

export default GetAction;
