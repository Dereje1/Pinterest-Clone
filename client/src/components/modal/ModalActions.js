import React from 'react';
import PropTypes from 'prop-types';
import IconButton from '@mui/material/IconButton';
import PushPinIcon from '@mui/icons-material/PushPin';
import PushPinOutlinedIcon from '@mui/icons-material/PushPinOutlined';
import DeleteForeverIcon from '@mui/icons-material/DeleteForever';
import DeleteIcon from '@mui/icons-material/Delete';
import Tooltip from '@mui/material/Tooltip';

const GetAction = ({
  element, pinImage, deletePin, reset,
}) => {
  if (!pinImage) { // means called from profile page
    return (
      <IconButton
        aria-label="settings"
        style={{ margin: '1vh' }}
        onClick={() => {
          deletePin(element);
          reset();
        }}
        onMouseDown={e => e.preventDefault()}
      >
        {
          element.owns
            ? (
              <Tooltip title="Permanently delete image" placement="left">
                <DeleteForeverIcon style={{ fontSize: '2em', color: '#d12929' }} />
              </Tooltip>
            ) : (
              <Tooltip title="Unpin image" placement="left">
                <DeleteIcon style={{ fontSize: '2em', color: '#d12929' }} />
              </Tooltip>
            )
        }
      </IconButton>
    );
  }
  // finds the status of image to determine what kind of button to place on pic
  if (element.hasSaved || element.owns) { // If the user has already saved this pin
    return (
      <IconButton aria-label="settings" style={{ margin: '1vh' }} disableRipple>
        <Tooltip title={element.owns ? 'You own this image' : 'You pinned this image'} placement="left">
          <PushPinIcon style={{ fontSize: '2em', color: element.owns ? '#c50000' : '#3a1c1cde' }} />
        </Tooltip>
      </IconButton>
    ); // no button
  }
  // user has not saved this pin show save button
  return (
    <IconButton
      aria-label="settings"
      style={{ margin: '1vh' }}
      onClick={() => {
        pinImage(element);
        reset();
      }}
      onMouseDown={e => e.preventDefault()}
    >
      <Tooltip title="Pin image" placement="left">
        <PushPinOutlinedIcon style={{ fontSize: '2em' }} />
      </Tooltip>
    </IconButton>
  );
};

export default GetAction;

GetAction.defaultProps = {
  pinImage: null,
  deletePin: null,
};

GetAction.propTypes = {
  element: PropTypes.objectOf(PropTypes.any).isRequired,
  // what type of button to place on pic/thumbnail executed by caller
  pinImage: PropTypes.func,
  deletePin: PropTypes.func,
  reset: PropTypes.func.isRequired,
};
