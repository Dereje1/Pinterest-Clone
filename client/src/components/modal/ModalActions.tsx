import React from 'react';
import IconButton from '@mui/material/IconButton';
import PushPinIcon from '@mui/icons-material/PushPin';
import PushPinOutlinedIcon from '@mui/icons-material/PushPinOutlined';
import DeleteForeverIcon from '@mui/icons-material/DeleteForever';
import Tooltip from '@mui/material/Tooltip';
import { PinType } from '../../interfaces';

interface GetActionProps{
  element: PinType
  reset: (_: React.SyntheticEvent, forceClose?: boolean) => void
  pinImage: (pin: PinType) => void
  deletePin: (pin: PinType) => void
}

function GetAction({
  element,
  pinImage,
  deletePin,
  reset,
}: GetActionProps) {
  if (!pinImage) { // means called from profile page
    return (
      <IconButton
        aria-label="settings"
        style={{ margin: '1vh' }}
        onClick={(e) => {
          deletePin(element);
          reset(e);
        }}
        onMouseDown={(e) => e.preventDefault()}
      >
        {
          element.owns
            ? (
              <Tooltip title="Permanently delete image" placement="bottom">
                <DeleteForeverIcon style={{ fontSize: '2em', color: '#d12929' }} />
              </Tooltip>
            ) : (
              <Tooltip title="Unpin image" placement="bottom">
                <PushPinIcon style={{ fontSize: '2em', color: '#3a1c1cde' }} />
              </Tooltip>
            )
        }
      </IconButton>
    );
  }

  // finds the status of image to determine what kind of pin to place on image
  if (element.hasSaved || element.owns) { // If the user has already saved this pin
    return (
      <IconButton
        aria-label="settings"
        style={{ margin: '1vh' }}
        disableRipple={element.owns}
        onClick={() => {
          if (element.hasSaved) {
            pinImage(element);
          }
        }}
        onMouseDown={(e) => e.preventDefault()}
      >
        <Tooltip title={element.owns ? 'You own this image' : 'Unpin this image'} placement="bottom">
          <PushPinIcon style={{ fontSize: '2em', color: element.owns ? '#c50000' : '#3a1c1cde' }} />
        </Tooltip>
      </IconButton>
    ); // no button
  }
  // user has not saved this pin show outlined pin
  return (
    <IconButton
      aria-label="settings"
      style={{ margin: '1vh' }}
      onClick={() => {
        pinImage(element);
      }}
      onMouseDown={(e) => e.preventDefault()}
    >
      <Tooltip title="Pin image" placement="bottom">
        <PushPinOutlinedIcon style={{ fontSize: '2em' }} />
      </Tooltip>
    </IconButton>
  );
}

export default GetAction;
