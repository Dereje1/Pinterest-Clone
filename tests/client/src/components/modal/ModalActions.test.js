import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ModalActions from '../../../../../client/src/components/modal/ModalActions';

describe('Handling zoomed image action buttons', () => {
  let props;
  beforeEach(() => {
    props = {
      element: {
        hasSaved: false,
        owns: false,
      },
      pinImage: jest.fn(),
      deletePin: jest.fn(),
      reset: jest.fn(),
    };
  });

  afterEach(() => {
    props = null;
  });

  test('will render the delete button for owners', () => {
    const updatedProps = {
      ...props,
      element: {
        hasSaved: false,
        owns: true,
      },
      pinImage: null,
    };
    const preventDefault = jest.fn();
    const wrapper = shallow(<ModalActions {...updatedProps} />);
    const iconButton = wrapper.find('ForwardRef(IconButton)');
    iconButton.props().onClick();
    iconButton.props().onMouseDown({ preventDefault });
    expect(props.deletePin).toHaveBeenCalledWith({
      hasSaved: false,
      owns: true,
    });
    expect(preventDefault).toHaveBeenCalled();
    expect(props.reset).toHaveBeenCalled();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will render the delete button for pinned images', () => {
    const updatedProps = {
      ...props,
      element: {
        hasSaved: false,
        owns: false,
      },
      pinImage: null,
    };
    const wrapper = shallow(<ModalActions {...updatedProps} />);
    const iconButton = wrapper.find('ForwardRef(IconButton)');
    iconButton.props().onClick();
    expect(props.deletePin).toHaveBeenCalledWith({
      hasSaved: false,
      owns: false,
    });
    expect(props.reset).toHaveBeenCalled();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will render the unpin button for already pinned images', () => {
    const updatedProps = {
      ...props,
      element: {
        hasSaved: true,
        owns: false,
      },
    };

    const wrapper = shallow(<ModalActions {...updatedProps} />);
    const preventDefault = jest.fn();
    const iconButton = wrapper.find('ForwardRef(IconButton)');
    iconButton.props().onClick();
    iconButton.props().onMouseDown({ preventDefault });
    expect(props.pinImage).toHaveBeenCalledWith({
      hasSaved: true,
      owns: false,
    });
    expect(preventDefault).toHaveBeenCalled();
    expect(props.reset).not.toHaveBeenCalled();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will render the disabled pin button for already owned images', () => {
    const updatedProps = {
      ...props,
      element: {
        hasSaved: false,
        owns: true,
      },
    };
    const wrapper = shallow(<ModalActions {...updatedProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });


  test('will render the pin button for unpinned images', () => {
    const preventDefault = jest.fn();
    const wrapper = shallow(<ModalActions {...props} />);
    const iconButton = wrapper.find('ForwardRef(IconButton)');
    iconButton.props().onClick();
    iconButton.props().onMouseDown({ preventDefault });
    expect(props.pinImage).toHaveBeenCalledWith({
      hasSaved: false,
      owns: false,
    });
    expect(preventDefault).toHaveBeenCalled();
    expect(props.reset).not.toHaveBeenCalled();
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
