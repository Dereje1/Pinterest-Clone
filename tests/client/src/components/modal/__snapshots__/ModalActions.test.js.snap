// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handling zoomed image action buttons will render the delete button for owners 1`] = `
<ForwardRef(IconButton)
  aria-label="settings"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "margin": "1vh",
    }
  }
>
  <ForwardRef(Tooltip)
    placement="bottom"
    title="Permanently delete image"
  >
    <Memo(ForwardRef(DeleteForeverIcon))
      style={
        Object {
          "color": "#d12929",
          "fontSize": "2em",
        }
      }
    />
  </ForwardRef(Tooltip)>
</ForwardRef(IconButton)>
`;

exports[`Handling zoomed image action buttons will render the delete button for pinned images 1`] = `
<ForwardRef(IconButton)
  aria-label="settings"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "margin": "1vh",
    }
  }
>
  <ForwardRef(Tooltip)
    placement="bottom"
    title="Unpin image"
  >
    <Memo(ForwardRef(PushPinIcon))
      style={
        Object {
          "color": "#3a1c1cde",
          "fontSize": "2em",
        }
      }
    />
  </ForwardRef(Tooltip)>
</ForwardRef(IconButton)>
`;

exports[`Handling zoomed image action buttons will render the disabled pin button for already owned images 1`] = `
<ForwardRef(IconButton)
  aria-label="settings"
  disableRipple={true}
  onClick={null}
  onMouseDown={[Function]}
  style={
    Object {
      "margin": "1vh",
    }
  }
>
  <ForwardRef(Tooltip)
    placement="bottom"
    title="You own this image"
  >
    <Memo(ForwardRef(PushPinIcon))
      style={
        Object {
          "color": "#c50000",
          "fontSize": "2em",
        }
      }
    />
  </ForwardRef(Tooltip)>
</ForwardRef(IconButton)>
`;

exports[`Handling zoomed image action buttons will render the pin button for unpinned images 1`] = `
<ForwardRef(IconButton)
  aria-label="settings"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "margin": "1vh",
    }
  }
>
  <ForwardRef(Tooltip)
    placement="bottom"
    title="Pin image"
  >
    <Memo(ForwardRef(PushPinOutlinedIcon))
      style={
        Object {
          "fontSize": "2em",
        }
      }
    />
  </ForwardRef(Tooltip)>
</ForwardRef(IconButton)>
`;

exports[`Handling zoomed image action buttons will render the unpin button for already pinned images 1`] = `
<ForwardRef(IconButton)
  aria-label="settings"
  disableRipple={false}
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "margin": "1vh",
    }
  }
>
  <ForwardRef(Tooltip)
    placement="bottom"
    title="Unpin this image"
  >
    <Memo(ForwardRef(PushPinIcon))
      style={
        Object {
          "color": "#3a1c1cde",
          "fontSize": "2em",
        }
      }
    />
  </ForwardRef(Tooltip)>
</ForwardRef(IconButton)>
`;
